tower-http/src/set_response_header.rs

```diff
@@ -0,0 +1,457 @@
+use tower_service::Service;
+
+/// Layer that applies [`SetResponseHeader`] which adds a response header.
+///
+/// See [`SetResponseHeader`] for more details.
+pub struct SetResponseHeaderLayer<M, T> {
+    header_name: HeaderName,
+    make: M,
+    mode: InsertHeaderMode,
+    _marker: PhantomData<fn() -> T>,
+}
+
+impl<M, T> fmt::Debug for SetResponseHeaderLayer<M, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("SetResponseHeaderLayer")
+            .field("header_name", &self.header_name)
+            .field("mode", &self.mode)
+            .field("make", &std::any::type_name::<M>())
+            .finish()
+    }
+}
+
+impl<M, T> SetResponseHeaderLayer<M, T> {
+    /// Create a new [`SetResponseHeaderLayer`].
+    pub fn new(header_name: HeaderName, make: M) -> Self
+    where
+        M: MakeHeaderValue<T>,
+    {
+        Self {
+            make,
+            header_name,
+            mode: InsertHeaderMode::OverrideExisting,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Set which mode to use when inserting the header.
+    ///
+    /// Defaults to [`InsertHeaderMode::OverrideExisting`].
+    pub fn mode(mut self, mode: InsertHeaderMode) -> Self {
+        self.mode = mode;
+        self
+    }
+}
+
+/// The mode to use when inserting a header into a request or response.
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub enum InsertHeaderMode {
+    /// Insert the header, overriding any previous values the header might have.
+    OverrideExisting,
+    /// Append the header and keep any previous values.
+    Append,
+    /// Insert the header only if it is not already present.
+    SkipIfPresent,
+}
+
+impl<T, S, M> Layer<S> for SetResponseHeaderLayer<M, T>
+where
+    M: MakeHeaderValue<T> + Clone,
+{
+    type Service = SetResponseHeader<S, M>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        SetResponseHeader {
+            inner,
+            header_name: self.header_name.clone(),
+            make: self.make.clone(),
+            mode: self.mode.clone(),
+        }
+    }
+}
+
+impl<M, T> Clone for SetResponseHeaderLayer<M, T>
+where
+    M: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            make: self.make.clone(),
+            header_name: self.header_name.clone(),
+            mode: self.mode.clone(),
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// Middleware that sets a header on the response.
+#[derive(Clone)]
+pub struct SetResponseHeader<S, M> {
+    inner: S,
+    header_name: HeaderName,
+    make: M,
+    mode: InsertHeaderMode,
+}
+
+impl<S, M> SetResponseHeader<S, M> {
+    /// Create a new [`SetResponseHeader`].
+    pub fn new(inner: S, header_name: HeaderName, make: M) -> Self {
+        Self {
+            inner,
+            header_name,
+            make,
+            mode: InsertHeaderMode::OverrideExisting,
+        }
+    }
+
+    /// Set which mode to use when inserting the header.
+    ///
```

Comment:
as above, maybe something like
```suggestion
    /// Configures how existing header values are handled.
    ///
    /// This takes an [`InsertHeaderMode`] which configures the service's
    /// behavior when other values have previously been set for the same
    /// header. The available options are:
    ///
    /// - `InsertHeaderMode::OverrideExisting` (the default): if a previous
    ///   value exists for the same header, it is removed and replaced with
    ///   the new header value.
    /// - `InsertHeaderMode::SkipIfPresent`: if a previous value exists for
    ///   the header, the new value is not inserted.
    /// - `InsertHeaderMode::Append`: the new header is always added,
    ///   preserving any existing values. If previous values exist, the header
    ///   will have multiple values.
```